### ISSUE TITLE
Add support for Eufy X10 Pro Omni

### ISSUE THREAD

**Issue by RickSisco** [CodeFoodPixels/robovac#68]
Are there plans to add support for the X10 Pro Omni vac?

**Comment by CodeFoodPixels** [comment]
As part of the work I'm doing on the `better-dps` branch, I'm making it so that devices with different commands can be supported. I'd need people to contribute/test the commands for the vacuums though.

**Comment by russilker** [comment]
Happy to be a tester for my X10 Pro Omni as needed.

**Comment by RickSisco** [comment]
If step by step instructions are supplied to test the commands, I would be
willing to help test on my X10 Pro Omni

**Comment by damfuu** [comment]
I am happy to help as well.

**Comment by RoadXY** [comment]
Does that mean there might be some light at the horizon for: https://github.com/CodeFoodPixels/robovac/issues/46 ass well? 😎

**Comment by CodeFoodPixels** [comment]
Yep, all of these are related to #28

**Comment by daschick111** [comment]
Count me in as a tester for the #46 issue

**Comment by gmagnus1** [comment]
Happy to help here as well.

**Comment by terabyte128** [comment]
Hello, I've been doing some work with my X10 to try and figure out where the setup process stalls. So far, I've noticed a couple of things:
1) The Eufy API endpoint to list devices seems to have changed.  When hitting `api/v1/device/list/devices-and-groups`, I don't see anything in the `items` array.

```json
{
  "res_code": 1,
  "message": "",
  "items": [],
  "global_config": {
    "enabled_multi_color_modes": [
      0,
      1,
      2,
      3
    ]
  }
}
```

I was able to MITM the traffic from the Eufy Clean app, though, and it looks like it instead makes a request to `/api/v1/device/v2`, which actually gives the device info:

```json
{
  "res_code": 1,
  "message": "",
  "devices": [
    {
      "id": "<snipped>",
      "sn": "",
      "name": "X10 Pro Omni",
      "alias_name": "Dustin",
      "bluetooth": null,
      "wifi": {
        "mac": "<snipped>",
        "wifi_ssid": "",
        "lan_ip_addr": ""
      },
      "product": {
        "id": "<uuid>",
        "name": "X10 Pro Omni",
        "region": "[{\"regions\":[\"ALL\"],\"date\":\"2022-09-01 20:22:49\"}]",
        "default_name": "RoboVac",
        "icon_url": "https://d1teb1w17vl5yo.cloudfront.net/eufyhome/products/T2182_addProduct.png",
        "category": "Home",
        "appliance": "Cleaning",
        "connect_type": 2,
        "description": "T2182 eufy RoboVac L80, connected via EufyHome",
        "product_code": "T2351",
        "wifi_ssid_prefix": "eufy Clean X10 Pro Omni",
        "wifi_ssid_prefix_full": "eufy Clean X10 Pro Omni-",
        "index": 1,
        "create_time": 1640585895,
        "update_time": 1707192290,
        "is_show": false,
        "tuya_pid": "<id>",
        "app_ble_ssid_prefix": "eufy Clean X10 Pro Omni-",
        "device_ble_ssid_prefix": "eufy Clean X10 Omni-",
        "wifi_ssid_prefix_list": [
          "eufy Clean X10 Omni-",
          "eufy Clean X10 Pro Omni-"
        ],
        "device_ble_ssid_prefix_list": [
          "eufy Clean X10 Omni-",
          "eufy Clean X10 Pro Omni-"
        ]
      },
      "user_id": "<uuid>",
      "owner_info": null,
      "home_id": "<uuid>",
      "home_name": "",
      "room_id": "<uuid>",
      "room_name": "Default Room",
      "connect_type": 2,
      "grant_by": 0,
      "software_version": "",
      "index": 0,
      "device_key": "",
      "create_time": 1711493109,
      "update_time": 1711493549,
      "hardware_version": "",
      "scale_type": "",
      "local_code": "",
      "needUpdate": false,
      "setting": {
        "id": "",
        "device_id": "",
        "is_default": true
      },
      "update_packages": []
    }
  ],
  "groups": []
}
```

Next, I connected to the Tuya API and tried to retrieve the device with:

```python
tuya.get_device("<device id>")
```
with both `devices[0]['id']` and `devices[0]['product']['id']`

but I've only gotten back
```
{'t': 1711599563374, 'success': False, 'errorCode': 'PERMISSION_DENIED', 'status': 'error', 'errorMsg': 'PERMISSION_DENIED'}
```

and I'm not sure what to do from here. It looks like the response from Tuya is pretty generic, so I possibly messed up something about the parameters.

This is the code that I've been playing around with, for reference:

```python
import requests

from tuyawebapi import TuyaAPISession

login = requests.post(
    "https://home-api.eufylife.com/v1/user/email/login",
    json={
        "client_id": "eufyhome-app",
        "client_secret": "GQCpr9dSp3uQpsOMgJ4xQ",
        "email": "<snipped>",
        "password": "<snipped>",
    },
)
login.raise_for_status()
login_data = login.json()

access_token = login_data["access_token"]
refresh_token = login_data["refresh_token"]
base = login_data["user_info"]["request_host"]

phone_code = login_data["user_info"]["phone_code"]
country = login_data["user_info"]["country"]
timezone = login_data["user_info"]["timezone"]
user_id = login_data["user_info"]["id"]

session = requests.Session()
session.headers["token"] = access_token

devices = session.get(f"{base}/v1/device/v2", headers={"category": "Home"})
devices.raise_for_status()
devices_data = devices.json()

tuya = TuyaAPISession(
    username=f"eh-{user_id}",
    region="AZ",
    timezone=timezone,
    phone_code=phone_code,
)
```

I also port-scanned the vacuum, in case that's helpful, but didn't find any of the known local Tuya ports to be open:
```
Nmap scan report for <snipped>
Host is up (0.010s latency).
Not shown: 1000 closed tcp ports (conn-refused)
PORT     STATE SERVICE
9668/tcp open  tec5-sdctp

Nmap done: 1 IP address (1 host up) scanned in 2.08 seconds
```
Unfortunately I wasn't able to glean that much other useful information from the app either. Starting a cleaning cycle triggers a bunch of requests to `log.eufylife.com/push_log_hdfs` and `log.eufylife.com/push_log_es` but nothing that is obviously the start signal. I'm happy to share more details about the flows. Note that I was only snooping on HTTP traffic, though, so if the app communicated locally with the vacuum over something else, I wouldn't have seen it.

**Comment by CodeFoodPixels** [comment]
That's some amazing work @terabyte128! Looks like the eufy endpoint you found works for older devices too, so I'm going to switch over to using that.

**Comment by terabyte128** [comment]
Thanks! I also noticed that the vacuum is sending out broadcast UDP packets from port **9667** (not 6666/6667), so that may also be part of the puzzle. Unfortunately I'm not sure how to decrypt them.

![image](https://github.com/CodeFoodPixels/robovac/assets/1189703/d571e910-cb67-4ae5-b4fe-9a050c496620)

(I doubt there's anything personal in the payload, but I'm gonna avoid posting it publicly – if you want it, let me know.)

**Comment by RickSisco** [comment]
Great work Sam. Look forward to seeing the progress Luke. :)

**Comment by terabyte128** [comment]
@CodeFoodPixels do you know how the username `eh-{user_id}` was deduced? (from [here](https://github.com/CodeFoodPixels/robovac/blob/main/custom_components/robovac/config_flow.py#L130)). I think there might be something wrong with how I'm using the Tuya API. 

Entering a random ID and calling `list_homes()` gives more or less the same answer as when I use my real ID, though, so it's hard to tell if my login is correct or not.

For example:

```python
tuya = TuyaAPISession(
    username=f"FAKEUSERID"
    region="AZ",
    timezone=timezone,
    phone_code=phone_code,
)
```

gives the response:

```python
[{'geoName': '', 'rooms': [], 'gmtModified': 1711668412, 'role': 2, 'gid': 192829708, 'groupId': 192829708, 'displayOrder': 1, 'admin': True, 'dealStatus': 2, 'gmtCreate': 1711668412, 'ownerId': '192829708', 'uid': 'az1711668412228EOl6W', 'groupUserId': 157327173, 'background': '', 'name': 'My Home', 'id': 147452158, 'status': True}]
```

while making the same request with `eh-{user_id}` gives the exact same response shape, just with different IDs etc.

**Comment by CodeFoodPixels** [comment]
> Thanks! I also noticed that the vacuum is sending out broadcast UDP packets to port **9667** (not 6666/6667), so that may also be part of the puzzle.

Good spot! Doing a quick search through a decompiled version of the eufy app, `9667` and `9668` seem to be used in place of `6667` and  `6668`.

> @CodeFoodPixels do you know how the username `eh-{user_id}` was deduced?

It was deduced before I had anything to do with this, but again, looking at the app code, it's right. Did you use an actual ID or literally `FAKEUSERID`?

**Comment by terabyte128** [comment]
I literally 

> It was deduced before I had anything to do with this, but again, looking at the app code, it's right. Did you use an actual ID or literally `FAKEUSERID`?

I literally used `FAKEUSERID`, lol. Maybe that's a way for them to defend against attackers trying to enumerate valid user IDs? Especially since they don't seem to require any sort of client secret.

**Comment by terabyte128** [comment]
I didn't notice any requests from the app going directly to Tuya servers, but I wasn't specifically looking for that – I'll take another look.

**Comment by CodeFoodPixels** [comment]
I think the contact with the tuya servers is mainly to get keys etc, possibly also when you're not on the same network as the vacuum.

**Comment by CodeFoodPixels** [comment]
@terabyte128 If you want some code to play around with, you can use this as a base: https://github.com/CodeFoodPixels/robovac-schema

**Comment by terabyte128** [comment]
I see, I wonder if it only contacts the Tuya server for local keys on initial setup. Thanks for the link – I’ll poke at it more this weekend!

**Comment by terabyte128** [comment]
I was able to capture some more information! I found that the app was making requests to `/api.json`, but on a different server than was encoded in `tuyawebapi.py` – in my case, `https://a1-us.iotbing.com/api.json`. Not sure if that makes a difference.

Perhaps more importantly, this is what the request format looks like:
```
Host:             a1-us.iotbing.com
Content-Type:     application/x-www-form-urlencoded
User-Agent:       EufyHome/20 CFNetwork/1474 Darwin/23.0.0
Connection:       keep-alive
Accept:           */*
Accept-Language:  en-US,en;q=0.9
Content-Length:   1555
Accept-Encoding:  gzip, deflate, br
Cache-Control:    no-cache
==== FORM DATA ===
lang:              en
bizData:           {"nd":1,"customDomainSupport":"1"}
deviceId:          83C29192-633C-4829-8CC0-CB30B65E6167
et:                0.0.2
osSystem:          17.0
bundleId:          com.eufylife.EufyHome
time:              1711840368
lon:               0
channel:           sdk
nd:                1
appVersion:        3.1.0
ttid:              appstore_d
os:                IOS
v:                 1.0
sid:               <snipped>
sign:              <snipped>
platform:          iPhone 11 Pro
postData:          USQUo1wT3zC1Mggnw4Dq8GiAp+pAFKpnkC1Bfiz53pphbXbtpMH2two1Bt/d9foJlYJXqxCUBe23WxRQy2w56kfAXQ18GEWFRK+u5TSULsgRJRW6q1ZOYkKBAU4Is++NMseQ2kMUXr/T3f+UvA4JeD+6xcoTqwrkd3zOa7Sbw878YMxkYklrFaaKOvK8lB+KL2h
zkOx/5rG34U8DQBwMv+aoVlMCZldQVNtd+5ahJhBweBIQ9gLZZkX90JarFy4F6dt4a+R3jwsZcv7L61ly3gJhopchbtzlWv0ETtx5F0NAN2MclLuyMvfoke4vR2XlihNzvdbZfVqEpzhvvMqDw2LFwVXy1Pfw8SHVRzuQCixxmI4lNJx/Jz3Ay++XKMQFTGPumbU7ryn4N+3ptsxGMIN/7
UP5WS3P5NQL50XvrjPsowkPAuMiCwnGw/Hfj29uIi2IobRL8MBspo6iM6pNCNZjexCFDJB4BkS4Vxk8YbqfIbCMmuHQJHoIE/qmico5q3684aVhNnI+fR/nRVRExHCkvl33+mpK8pSqVTRTyficFrpnHBbLnC1HdBpniayunxI5XNHjLfW3Pooj5Vp6DXE1t6pvuxqBmqpftQfQyFpKjwL
UAX2xccHevYnEMInLPxWuThci6yvQdU5yJEZOssfHpUpLnzDIcLTB8i5jc6PxZAnyi/FJPAiH5SUEhr0poIr0U4QCHS5AakCsNtz2p7bcwkrDQ9BwzMX7RoyxJKjDhc6nCyj9B3Pk6vXOyzx3WjEka0fSGfRRP+VlUY2z2sK2hRdF3ge6WH9IgesVmBokh2XQGggBGuSwab37PmgR6Iava
9rfU6bWO8qQX4kJGw==
uid:               az1711493114821bjS7C
sdkVersion:        4.3.0
timeZoneId:        America/Los_Angeles
requestId:         <snipped>
lat:               0
gid:               192606234
clientId:          <snipped>
deviceCoreVersion: 5.3.0
a:                 smartlife.m.api.batch.invoke
cp:                gzip
```

The most interesting thing here is that `postData` appears to be encrypted, which is different than what `tuyawebapi.py` does. Looking at [this link](https://github.com/codetheweb/tuyapi/issues/20#issuecomment-984693061) supports this idea as well, and it also offers a function that can apparently perform the encryption. However, I'm not sure what how to obtain `tuya_bmpkey` or `tuya_appsecret`.

**Comment by terabyte128** [comment]
I have another progress update. Long story short, from what I can tell it appears that the Eufy Clean app doesn't communicate directly with the vacuum at all. Instead, it sends all commands through a Eufy MQTT server (in my case, `aiot-mqtt-us.anker.com:8883`). The MQTT server uses TLS.

* The username for the MQTT server appears to be `{user id}-eufy_home`, with no password.
* The topic in my case is `cmd/eufy_home/T2351/{vacuum ID}/req` (not sure what T2351 represents)
* The commands themselves look like:

```json
{
    "head": {
        "client_id": "android-eufy_home-{user id}-eufy_android_Android SDK built for arm64_{some random hex digits, maybe a version number?}",
        "cmd": 65537,
        "cmd_status": 2,
        "msg_seq": 1,
        "seed": "",
        "sess_id": "android-eufy_home-{user id}-eufy_android_Android SDK built for arm64_{same hex digits as above}",
        "sign_code": 0,
        "timestamp": 1712295579241,
        "version": "1.0.0.1"
    },
    "payload": "{\"account_id\":\"{account id}\",\"data\":{\"153\":\"BwjpnJTm6jE\\u003d\"},\"device_sn\":\"{vacuum id}\",\"protocol\":2,\"t\":1712295579241}"
}
```
`payload['data']` appears to be base64-encoded, and happily, not encrypted. The bytes seem to correspond exactly to the action performed, and don't change. For example (first digit is the key in the `data` dict):

```
154 16,10,14,10,2,8,2,26,0,34,2,8,1,42,2,8,1  # turn on 'auto' mode
154 14,10,12,10,2,8,2,26,0,34,2,8,1,42,0  # turn off 'auto' mode
```

Unfortunately I haven't been able to send out commands on my own yet. I suspect there might be another piece of auth that I'm missing before I can publish commands to the topic, and/or it might want a client certificate.

**Comment by RickSisco** [comment]
Thanks for the update. Glad to see you finding progress.

**Comment by RoadXY** [comment]
T2351 is the model number of your vacuum.
T2351 = X10 Pro omni

https://support.eufy.com/s/article/T2351-EU-DOC

**Comment by terabyte128** [comment]
I was able to connect to the MQTT server and inspect all the messages sent to the relevant topics. I could also publish on the topics and cause the vacuum to perform actions. The MQTT topics are:
* `cmd/eufy_home/T2351/{VACUUM_SERIAL}/req`: requests send from the app to the vacuum
* `cmd/eufy_home/T2351/{VACUUM_SERIAL}/res` responses from the vacuum

Below is the snippet that I used. The important pieces are:

* The MQTT user is `{USER_ID}-eufy_home`
* You can only subscribe to the topics namespaced under `cmd/eufy_home/T2351/{VACUUM_SERIAL}`
* The MQTT server **does** require the client to present a certificate. I got it by instrumenting the Android Eufy Clean app. I'm not sure I should post the private key here, but you can pretty easily get it yourself. Here's the Frida snippet I used:

```javascript
Java.perform(function () {
  let MqttConnectionNew = Java.use("com.anker.aiot_sdk.mqtt.MqttConnectionNew");
  MqttConnectionNew["s"].implementation = function (domainMqttBean) {
    console.log(
      `MqttConnectionNew.s is called: domainMqttBean=${domainMqttBean}`,
    );
    let result = this["s"](domainMqttBean);
    console.log(`MqttConnectionNew.s result=${result}`);
    return result;
  };
});
```

and the code I wrote to connect to the MQTT server:

```python
import json
from base64 import b64decode

import paho.mqtt.client as mqtt
from paho.mqtt.enums import CallbackAPIVersion

# the following can be retrieved using existing scripts against Eufy's API
USER_ID = "<snipped>"
VACUUM_SERIAL = "<snipped>"

client = mqtt.Client(
    CallbackAPIVersion.VERSION2,
    client_id=(
        f"android-eufy_home-{USER_ID}-eufy_android_Android"
        f" SDK built for arm64_{USER_ID}"
    ),
)
client.tls_set(
    certfile="./certs/certificate.pem",
    keyfile="./certs/private_key.pem",
)


def print_bytes(bs: bytes):
    ret = ""
    for b in bs:
        formatted = f"{b:02x} "
        ret += formatted

    return ret


def on_message(client: mqtt.Client, userdata, message: mqtt.MQTTMessage):
    topic = message.topic
    payload = json.loads(message.payload)

    if isinstance(payload["payload"], str):
        payload["payload"] = json.loads(payload["payload"])

    if "res" in topic:
        return
        print("<-- ", end="")
    else:
        print("--> ", end="")

    for k, v in payload["payload"]["data"].items():
        try:
            decoded = b64decode(v)
            print(f"{k}: {print_bytes(decoded)}")
            print(v)
        except:
            print(f"failed to decode {k}: {v}")

    # print(topic)
    # print(json.dumps(payload, indent=4))


client.on_message = on_message
client.username = f"{USER_ID}-eufy_home"
client.connect("aiot-mqtt-us.anker.com", 8883)

client.subscribe(f"cmd/eufy_home/T2351/{VACUUM_SERIAL}/#")

client.loop_forever()

```

The Python snippet just prints out the payloads of each message. Next step is figuring out what they actually mean. It seems like most payloads carry the _entire_ configuration of the vaccum's "mode", but since they're variable length, it's difficult to grok what bytes mean what. Here are some examples:

```
--> 154: 10 0a 0e 0a 02 08 02 1a 00 22 02 08 01 2a 02 08 01  # smart mode on
--> 154: 0e 0a 0c 0a 02 08 02 1a 00 22 02 08 01 2a 00  # smart mode off
--> 154: 10 0a 0e 0a 02 08 02 1a 02 08 02 22 02 08 01 2a 00 # fast, vacuum+mop, standard suction, medium water level
```

**Comment by terabyte128** [comment]
Eufy appears to be using some obscenely complicated protocol to generate the "set mode" payloads (seems like they're using an AST to build up the message like programming language syntax...why???) so rather than try and wade through all the code, I just brute-forced it. Below is the file that corresponds to the base64-encoded messages associated with each mode-tuple. Every "mode" message appears to be in the format of:

```json
{
    "154": "<mode tuple>"
}
```

[codes.csv](https://github.com/CodeFoodPixels/robovac/files/14953543/codes.csv)

**Comment by terabyte128** [comment]
In a nicer format: https://gist.github.com/terabyte128/0598dcb735ec73842dfc5d204d968320

**Comment by lukemorrigan-pp** [comment]
Are they encoded as protobufs? That was the case for another device: https://github.com/CodeFoodPixels/robovac/issues/40#issuecomment-2027798578

**Comment by CodeFoodPixels** [comment]
Are they encoded as protobufs? That was the case for another device: https://github.com/CodeFoodPixels/robovac/issues/40#issuecomment-2027798578

**Comment by terabyte128** [comment]
Hmm, could be, I'll dig into that.

**Comment by samonthenetuk** [comment]
Has anyone got this working with their x10 pro? I just got my new hoover yesterday and cannot get it work.

**Comment by apeschel780** [comment]
Home assistant Integration for the Eufy x10 is still in progress, but if you only want to start a cleaning cycle with a Lovelace card or automation, you can integrate your x10 with the Eufy skill in Alexa and create a cleaning routine in Alexa as well. In home assistant you can use the Alexa media player integration to start your routine. Here is my example:

service: media_player.play_media
target:
  entity_id: media_player.echo_im_bad
data:
  media_content_id: staubsaugen
  media_content_type: routine
metadata: {}

**Comment by Antopeto** [comment]
I am happy to be a guinea pig but i am not hot on coding so clear instructions will help me a lot!

**Comment by terabyte128** [comment]
I am still working on this but other things have taken priority recently, unfortunately.

**Comment by terabyte128** [comment]
@CodeFoodPixels unfortunately they don't seem to be protobufs, I don't see a consistent decoding using [protoscope](https://github.com/protocolbuffers/protoscope):

```shell
$ base64 -d <<< "DAoKCgAaACIAKgIIAQ==" | protoscope
1:EGROUP
1: {
  1: {}
  3: {}
  4: {}
  5: {1: 1}
}
~ $ base64 -d <<< "CgoICgAaACIAKgA=" | protoscope
`0a0a080a001a0022002a00`
```

**Comment by RickSisco** [comment]
We understand and appreciate your time. :)

**Comment by martijnpoppen** [comment]
@CodeFoodPixels @terabyte128 for what it's worth. I came accross the IObroker integration, and it seems he found a way to debug it. Not sure if it's helpful: https://github.com/TA2k/ioBroker.euhome/blob/main/main.js#L118

**Comment by terabyte128** [comment]
Ooh that is a good find! Looks very similar to what I’ve been experimenting with, but much more fleshed out.

**Comment by samonthenetuk** [comment]
> Home assistant Integration for the Eufy x10 is still in progress, but if you only want to start a cleaning cycle with a Lovelace card or automation, you can integrate your x10 with the Eufy skill in Alexa and create a cleaning routine in Alexa as well. In home assistant you can use the Alexa media player integration to start your routine. Here is my example:
> 
> service: media_player.play_media target: entity_id: media_player.echo_im_bad data: media_content_id: staubsaugen media_content_type: routine metadata: {}

I was hoping to integrate with Home Assistant so that I could add it to homekit.

**Comment by martijnpoppen** [comment]
@terabyte128 I tried also multiple things to decode the strings.
Seems like it's a Tuya specific encoding. (i thought i might be eufy encoding aswell).

But found this thread: https://github.com/codetheweb/tuyapi/discussions/440#discussioncomment-488832 not sure if this useful

**Comment by martijnpoppen** [comment]
@terabyte128 @CodeFoodPixels 
Had a breaktrough here ;)

So the values are Protobufs. But they are encoded delimited. (got some inspiration here: https://github.com/tuya/tuya-home-assistant/issues/495)

I made a setup in JS (sorry about that ;) )

```
const protobuf = require('protobufjs');

const main = async function () {
    const Fan = await getData('clean_param.proto', 'Fan', 'EAoOCgIIAhoAIgQIAhABKgA=');
    const MopMode = await getData('clean_param.proto', 'MopMode', 'EAoOCgIIAhoAIgQIAhABKgA=');
    const CleanCarpet = await getData('clean_param.proto', 'CleanCarpet', 'EAoOCgIIAhoAIgQIAhABKgA=');
    const CleanType = await getData('clean_param.proto', 'CleanType', 'EAoOCgIIAhoAIgQIAhABKgA=');
    const CleanExtent = await getData('clean_param.proto', 'CleanExtent', 'EAoOCgIIAhoAIgQIAhABKgA=');
    const CleanTimes = await getData('clean_param.proto', 'CleanTimes', 'EAoOCgIIAhoAIgQIAhABKgA=');

    console.log('Fan', Fan);
    console.log('MopMode', MopMode);
    console.log('CleanCarpet', CleanCarpet);
    console.log('CleanType', CleanType);
    console.log('CleanExtent', CleanExtent);
    console.log('CleanTimes', CleanTimes);
};

const getData = async function (proto, type, base64Value) {
    const root = await protobuf.load(proto);

    // Resolve the WorkStatus message type
    const WorkStatus = root.lookupType(type);

    // Decode the base64 value to a buffer
    const buffer = Buffer.from(base64Value, 'base64');

    // Decode the hex value
    //   const buffer = Buffer.from(hexValue, 'hex');
    const decodedMessage = WorkStatus.decodeDelimited(buffer);

    // Convert the decoded message to a plain JavaScript object
    const decodedObject = WorkStatus.toObject(decodedMessage, {
        longs: String, // Long objects will be converted to strings
        enums: String, // Enum values will be converted to strings
        bytes: String // Bytes will be converted to base64 strings
    });

    return decodedObject;
};

main();

```


This results in:

```
Fan { suction: 'TURBO' }
MopMode { level: 'HIGH' }
CleanCarpet { strategy: 'IGNORE' }
CleanType { value: 'SWEEP_AND_MOP' }
CleanExtent { value: 'QUICK' }
CleanTimes { autoClean: 2, selectRooms: 1, spotClean: 4 }
```


The Proto files I got from the decompiled APK.

**Comment by terabyte128** [comment]
@martijnpoppen very cool, would you mind giving me a hint about where you found those files in the APK? A path would be great :)

**Comment by martijnpoppen** [comment]
@terabyte128 resources/proto/cloud

**Comment by terabyte128** [comment]
Finally have some time to look at this again :)

It looks like even though they have the total length at the beginning of the message, only one message is ever sent as a time. I think the reason that you could interpret `EAoOCgIIAhoAIgQIAhABKgA=` as all those types is because it is actually a `CleanParamRequest`, which contains all of them:

```proto
message CleanParam {
    CleanType clean_type = 1;
    CleanCarpet clean_carpet = 2;
    CleanExtent clean_extent = 3;
    MopMode mop_mode = 4;
    Switch smart_mode_sw = 5; // 智能省心模式开关
    Fan fan = 6;  // 从 x10 项目开始使用这里的风机档位，旧项目使用单独 dp
    uint32 clean_times = 7; // 清扫次数，非 0 有效
}

message CleanParamRequest {
    CleanParam clean_param = 1;
    CleanParam area_clean_param = 2;
}
```

I was able to get this working in Python by just chopping off the first byte from the decoded base64 (because it's a `VARINT` and the lengths are never long enough to require more than one byte):

```python
from typing import Type, TypeVar

from google.protobuf.message import Message
from proto.cloud.clean_param_pb2 import CleanParamRequest
from proto.cloud.work_status_pb2 import WorkStatus
from base64 import b64decode

DP_MAP: dict[int, Type[Message]] = {
    153: WorkStatus,
    154: CleanParamRequest,
}

T = TypeVar("T", bound=Type[Message])


def decode(b64_data: str, to_type: T, has_length: bool = True) -> T:
    data = b64decode(b64_data)

    if has_length:
        data = data[1:]

    return to_type().FromString(data)


def encode(data: Message, has_length: bool = True) -> bytes:
    out = data.SerializeToString()

    if has_length:
        out = bytes([len(out)]) + out

    return out
```

**Comment by martijnpoppen** [comment]
@terabyte128 if you do your decode like i did in JS:

```js
    const decodedObject = protoLookupType.toObject(message, {
        longs: String, // Long objects will be converted to strings
        enums: String, // Enum values will be converted to strings
        bytes: String // Bytes will be converted to base64 strings
    });
    ```
    
    you'll get the value from the proto file aswell:
    
    instead of:
    ```js
    { cleanType: CleanType { value: 1 } }
    ```
    
    you'll get:

```js
{ cleanType: { value: 'MOP_ONLY' } }
```
    
    
I think you can to that with `MessageToDict`


Btw, did you get my mail with files? :)

**Comment by terabyte128** [comment]
Yes I did, thanks! 

Unfortunately the official Python library for protobufs is not as expressive as the one you used for JS, and I ran into issues with [this alternative](https://github.com/danielgtaylor/python-betterproto).

But printing out these objects does show the enum name, at least:

```
clean_param {
  clean_type {
    value: SWEEP_AND_MOP
  }
  clean_carpet {
  }
  clean_extent {
    value: NARROW
  }
  mop_mode {
    level: HIGH
  }
  smart_mode_sw {
  }
}
```

**Comment by terabyte128** [comment]
BTW: what does "DPs" actually stand for?

**Comment by martijnpoppen** [comment]
@terabyte128 Datapoints

**Comment by BarryOCathain** [comment]
Late to the party guys, but I've recently purchased an X10, so I can help out with debugging etc.

**Comment by cronner** [comment]
it's been a couple of months, any chance taht it's gonna be supported? could be really nice

**Comment by terabyte128** [comment]
Been busy with a variety of other things, but in the interest of having _something_ that can be worked on, I've published my progress so far at https://github.com/terabyte128/eufy-mqtt-vacuum.

This contains all the pieces to authenticate against the MQTT server, subscribe to status updates, and decode a subset of them from protobuf messages into Python dataclasses.

I'm not that familiar with how MQTT integrates into tools like home assistant, so I haven't been able to take much of a crack at that, but hopefully this is helpful for solving vacuum auth side of things. I'll probably keep poking at it as time allows, but hey, it's free labor – what do you expect :)

**Comment by terabyte128** [comment]
I figured it would make sense to split this into its own repo given that it's more-or-less an entirely new strategy for device authentication and is worth existing on its own so folks can use with tools over than HA.

**Comment by martijnpoppen** [comment]
Yep and also from my side my integration is done as well. Combining all different devices types. However thats for Homey and not HA so not useable here.

**Comment by leeandy1** [comment]
I have also just purchased an X10 so would be happy to help.

**Comment by leeandy1** [comment]
Just wondering if there was anything we could help with to get this working?

**Comment by terabyte128** [comment]
> Just wondering if there was anything we could help with to get this working?
You could open a PR ;)

**Comment by devPeete** [comment]
Just wondering if anyone started working on bringing this to home assistant. If not, I could think of having a look, I would just like to avoid redundant work.

**Comment by leeandy1** [comment]
@devPeete I am more than happy to help you with testing but I am not a coder. Unless others have already started I'm sure there are several users with X10's that would welcome your input.

**Comment by lirand84** [comment]
> Just wondering if anyone started working on bringing this to home assistant. If not, I could think of having a look, I would just like to avoid redundant work.

If necessary, I would also be happy to help with testing with my X10.

**Comment by cronner** [comment]
> Just wondering if anyone started working on bringing this to home assistant. If not, I could think of having a look, I would just like to avoid redundant work.

I can test on my x10 if you go forward :)

**Comment by aViN-187** [comment]
i'll add myself to the wishlist - waiting for the support of the x10 as well.

**Comment by charlesosj** [comment]
Happy to be another tested also, thanks for all your work!

**Comment by VlarpNL** [comment]
Same here. I would love to test support for the X10 Pro Omni!

**Comment by Nino6689** [comment]
> Just wondering if anyone started working on bringing this to home assistant. If not, I could think of having a look, I would just like to avoid redundant work.

I could also help with testing.

**Comment by Bas1978** [comment]
> > Just wondering if anyone started working on bringing this to home assistant. If not, I could think of having a look, I would just like to avoid redundant work.
> 
> 
> 
> I could also help with testing. 

As far as I know everyone is waiting for update for x10 pro Omni.

So if you can help us to update x10 pro Omni in home assistant yes please.

**Comment by devPeete** [comment]
Thanks for your offers for testing. I will have a look what is available the next days and let you know as soon as there is something testable.

**Comment by maximoei** [comment]
@terabyte128 your works looks super interesting. I've used parts of it to get my L60 SES to work via the existing Robovacs integration. Many things still missing though, maybe also interesting for X10? https://github.com/CodeFoodPixels/robovac/issues/85#issuecomment-2249935438

**Comment by cronner** [comment]
there is a working Homey app that does work with x10, can't something be used from that?

**Comment by devPeete** [comment]
@maximoei Thanks, I'll have a look at your work aswell.
@cronner Thanks for letting me know. I've found the sources for the homey app and will check for additional information

**Comment by martijnpoppen** [comment]
@devPeete i already shared my work in this thread -> https://github.com/CodeFoodPixels/robovac/issues/68#issuecomment-2104843008


Ps: I'm the developer of the eufy clean homey app

**Comment by martijnpoppen** [comment]
@devPeete If you need I can give you acces to my repo were I made the SDK for the homey app. It's still WIP though

**Comment by devPeete** [comment]
> @devPeete If you need I can give you acces to my repo were I made the SDK for the homey app. It's still WIP though

This could definetly help and would be great, Thanks!

**Comment by hoombar** [comment]
🖐️ also have an X10, and willing to help with any testing you need

**Comment by martijnpoppen** [comment]
@devPeete Invited you to the repo. Keep in mind it's WIP. For any questions, you can create a issue

**Comment by simplicityscott** [comment]
I too have an Omni Pro X10 and happy to assist with any testing.  I have noticed that even with the Eufy "skill" for Alexa...Alexa is not able to do much with the Omni Pro X10...for example..."Alexa, tell Eufy to vacuum the kitchen"...a capability I always had with iRobot...but I won't talk about my experiences with iRobot here.  :)

**Comment by Bas1978** [comment]
Is there any progress to add x10 Omni pro to home assistant?

**Comment by devPeete** [comment]
> Is there any progress to add x10 Omni pro to home assistant?

Still on it, I do not have the most time for it atm. However, I did not forget you! I am still looking at the existing source, I was already able to connect my own device with the Python Script. Hence, there was no major change by eufy that breaks the existing work. 
Still on question to all of you: Are you getting something by accessing http://<your_device_ip_here>/shell from your webbrowser?

**Comment by mattjgalloway** [comment]
> Still on question to all of you: Are you getting something by accessing http://<your_device_ip_here>/shell from your webbrowser?

That doesn't work for my X10 Pro Omni (or S1 Pro for that matter).

The only port open on my X10 Pro Omni is 9668.

**Comment by knackrack615** [comment]
Can confirm this also does nothing on my X10 Pro Omni (Connection refused)

**Comment by leeandy1** [comment]
It does nothing on mine either

**Comment by terabyte128** [comment]
@devPeete as far as my testing revealed, the X10 has no open HTTP ports and communicates solely via MQTT using Eufy’s broker.

**Comment by simplicityscott** [comment]
connection refused.

Sent from my iPhone

On Aug 14, 2024, at 12:17 PM, Matt Galloway ***@***.***> wrote:

﻿

Still on question to all of you: Are you getting something by accessing http://<your_device_ip_here>/shell from your webbrowser?

That doesn't work for my X10 Pro Omni (or S1 Pro for that matter).

The only port open on my X10 Pro Omni is 9668.

**Comment by Nino6689** [comment]
![image](https://github.com/user-attachments/assets/afd2c7d9-3e62-4e57-b793-93e675f85a32)

Here are my scan results. Only have 9668

**Comment by devPeete** [comment]
> @devPeete as far as my testing revealed, the X10 has no open HTTP ports and communicates solely via MQTT using Eufy’s broker.

Somehow, I know somebody with a device that has a shell and a webui on that port. I guess they forgot to disable debug access on that device. However, I did not find much that helped in reverse engineering the device.

**Comment by martijnpoppen** [comment]
@devPeete Did you see the invite for my repo ? The X10 and the S1 are fully supported there :)

**Comment by Cortexian** [comment]
Getting an X10 Omni tomorrow. If I can help in any way please let me know! Would love to get some tie ins to HA for scheduling purposes and some better voice activated cleaning of specific areas!

**Comment by devPeete** [comment]
> @devPeete Did you see the invite for my repo ? The X10 and the S1 are fully supported there :)

Yes I've seen. Did not yet looked at each and every part there. But so far it was helping quite much. Thank you!

**Comment by roan111** [comment]
Is there also support for the x8 pro ses?

**Comment by eatoff8** [comment]
Happy to help in any way with X10 testing

**Comment by rwijnhov** [comment]
subbed

**Comment by avvugt** [comment]
Another X10 Pro user. Happy to help testing

**Comment by Madelineqt** [comment]
I have an x10 pro too, subbing to this, can help with testing too

**Comment by sanohin** [comment]
I've got x10 as well, happy to test as well 😅

**Comment by cronner** [comment]
don't think it's happening?

**Comment by leeandy1** [comment]
I wouldn't get your hopes up. I think a couple of people have taken a look but other commitments are maybe taking preference. 

I did reach out to Eufy to see if they have any plans to make integration easier but their customer service were not very helpful.

**Comment by CodeFoodPixels** [comment]
I don't have an x10, so it's pretty difficult for me to do anything with
it. I'll happily accept a pull request.

**Comment by martijnpoppen** [comment]
@CodeFoodPixels as shared before in this topic I have successfully integrated the X10, S1, X9, X8 pro in the Homey integration 

I have a separate repo (written in TS) which I can share with you if you want.

Also shared earlier with @devPeete

**Comment by johnnieblows** [comment]
> @CodeFoodPixels as shared before in this topic I have successfully integrated the X10, S1, X9, X8 pro in the Homey integration
> 
> I have a separate repo (written in TS) which I can share with you if you want.
> 
> Also shared earlier with @devPeete
> 
> The repo contains of 3 ways of connecting
> 
> * local via tuyapi (doesn't contain the encrypted dps)
> * cloud via tuya cloud (this works for X9 / S1)
> * MQTT for the X10

@martijnpoppen would you be able to share your repo with me? 

Thanks

**Comment by martijnpoppen** [comment]
@johnnieblows sure!
Keep in mind it's not the perfect repo so might have some errors :)

**Comment by sanohin** [comment]
@martijnpoppen would you mind sharing access with me as well please? 

Thank you

**Comment by devPeete** [comment]
I am still on it. Have to admit that there is quite a lot besides this topic for me currently. However, I still have this topic on my list. Maybe it would be a good idea to sync in some way @sanohin @johnnieblows .

**Comment by leeandy1** [comment]
@devPeete that is great news, as they saying goes many hands make light work.

I would just like to say a massive thank you to all the developers who give up their time to make integrations like this for the community. I for one am truly grateful.

**Comment by martijnpoppen** [comment]
@sanohin Instead of adding you to the repo i decided to make the repo public 😊 

SDK: https://github.com/martijnpoppen/eufy-clean

Please note that it might not be super clear how it works. PR's are welcome. 🚀

To see how i Integrated it with my Homey app:

APP: https://github.com/martijnpoppen/com.eufylife.home (check main-device.js in ./drivers/)

And if you have a Homey: https://homey.app/nl-nl/app/com.eufylife.home/Eufy-Clean/

**Comment by todanator** [comment]
I might be missing something @martijnpoppen though it looks like the eufy-clean repo build depends on com.eufylife.home/lib/eufy-clean which I'm not seeing[ in that repo.](https://github.com/martijnpoppen/com.eufylife.home/tree/main/lib)

These commands specifically seem to be looking for something relative:

```
"copy-build": "cp -r ./build ../../com.eufylife.home/lib/eufy-clean/build && cp -r ./package.json ../../com.eufylife.home/lib/eufy-clean/ && cp -r ./tsconfig.json ../../com.eufylife.home/lib/eufy-clean/",
    "clear-build": "rimraf ../../com.eufylife.home/lib/eufy-clean/build",
 ```

**Comment by todanator** [comment]
@martijnpoppen this is great! I've got it wired up locally and I can at least see my vac. One thing - does your SDK have a way of vacuuming a specific room? I've yet to see the correct params in `setCleanParam` and `roomClean` doesn't seem to take any params. Maybe this is something accomplished with scene.proto? I do see room_id mentioned in there...

**Comment by todanator** [comment]
So far I've just been working on getting @martijnpoppen's sdk running locally with some vanilla node js to get familiar with the protobuffs he shared.

I'm working on a POC to fetch the room_id+name and then sending a clean request for that specific room. I haven't moved anything to python.

@martijnpoppen - for your decode util, how do you pick the `proto` and `type`? From the sdk logs I just see config.dps as a set of key/value pairs with nothing that denotes how they are delimited.

**Comment by martijnpoppen** [comment]
@todanator Yes you can add a room_id. But i haven't found a way to do so yet. next to that it was not really needed for my integration as there's no map to click a room. So didn't include it any further. 

What should work is clicking a room in the official app and then start it in the SDK, but that's just a extra step 😅

**Comment by martijnpoppen** [comment]
> @martijnpoppen - for your decode util, how do you pick the proto and type? From the sdk logs I just see config.dps as a set of key/value pairs with nothing that denotes how they are delimited.

Like this: https://github.com/martijnpoppen/eufy-clean/blob/main/src/controllers/SharedConnect.ts#L79

**Comment by martijnpoppen** [comment]
Btw there's also somebody else working on a integration: https://github.com/jimbo19922003/eufy-clean-ha-addon

**Comment by todanator** [comment]
Sorry if I'm missing it, but I'm not sure I get how you take the number plus the base64 encoded value and figure out which .proto file & message to use it with.

Maybe I'm thinking about it backward but I don't see how you figured out these existing mappings like [154 maps to Cleaning Parameters. 
](https://github.com/martijnpoppen/eufy-clean/blob/main/src/controllers/Base.ts#L22)

My current case is to fetch the map data, but I'm not sure which number to use for the encoded data.

```javascriptt
        const value = await encode('proto/cloud/multi_maps.proto', 'MultiMapsManageRequest', {
            method: 7,
            seq: 1  // Sequence number for tracking
        });
        
        // Send `value` to the vacuum's API and await the response.
        return await this.sendCommand({
            /* what number do I use here? */ : value
        })
```

**Comment by martijnpoppen** [comment]
@todanator i did by connecting a mqtt client and test it. But also checked in the decompiled android app.

With that info you can open the proto file and in there you can find the different parameters

**Comment by martijnpoppen** [comment]
@todanator I also checked the map request, but it seems like we only get mapIds but not really a way to retrieve the mapUrl from Tuya. 

Response of the maps request:

```
[
  {
    "key": "Result",
    "value": 2,
    "errCode": 107
  },
  {
    "key": "p2p.MapInfo",
    "releases": 2,
    "mapId": 107
  },
  {
    "key": "CompleteMaps",
    "completeMap": [
      {
        "mapId": 107
      }
    ]
  }
]
---------------------------------
Sending command to device AMP96Y0E03400419 {"account_id":"e5171d4dfbba1b99f4bd9978ab9d7392cd32e76e","data":{"172":"AggH"},"device_sn":"AMP96Y0E03400419","protocol":2,"t":1728286172571}
172
[
  {
    "key": "Result",
    "value": 2,
    "errCode": 107
  },
  {
    "key": "p2p.MapInfo",
    "releases": 2,
    "mapId": 107
  },
  {
    "key": "CompleteMaps",
    "completeMap": [
      {
        "mapId": 107
      }
    ]
  }
]
```

**Comment by martijnpoppen** [comment]
@todanator I also checked the map request, but it seems like we only get mapIds but not really a way to retrieve the mapUrl from Tuya. 

Response of the maps request:

```
[
  {
    "key": "Result",
    "value": 2,
    "errCode": 107
  },
  {
    "key": "p2p.MapInfo",
    "releases": 2,
    "mapId": 107
  },
  {
    "key": "CompleteMaps",
    "completeMap": [
      {
        "mapId": 107
      }
    ]
  }
]
```

**Comment by todanator** [comment]
Awesome! Thanks @martijnpoppen  - this unblocks me for now 💪🏻

**Comment by reynolpe** [comment]
I can help test if still needed. I have an x10 Omni pro. Would love to have it working with home assistant.

**Comment by jlightfo666** [comment]
I to would be happy to help test and give access to a vm of choice on the same network as the robot.

**Comment by todanator** [comment]
For anyone who wants a quick and dirty way to bring basic functionality into HA, you can use @martijnpoppen's SDK to just quickly start a Eufy Clean scene which and be any custom vacuum scenario.

For my own stop-gap solution, I just stood up nodejs endpoint that calls those scenes by number. HA has the ability to make REST calls. 

So my setup is HA Button -> HA REST call -> NodeJS endpoint -> Eufy Scene via SDK

**Comment by Rustymage** [comment]
> For anyone who wants a quick and dirty way to bring basic functionality into HA, you can use @martijnpoppen's SDK to just quickly start a Eufy Clean scene which and be any custom vacuum scenario.
> 
> For my own stop-gap solution, I just stood up nodejs endpoint that calls those scenes by number. HA has the ability to make REST calls.
> 
> So my setup is HA Button -> HA REST call -> NodeJS endpoint -> Eufy Scene via SDK

I wish I was smart enough to understand this.

**Comment by hades200082** [comment]
@todanator Do you have a docker compose for your REST solution?

**Comment by chutson92** [comment]
Has there been any additional progress made here?

**Comment by devPeete** [comment]
> Has there been any additional progress made here?

Sorry, not from my side currently. Maybe I will have some time in the upcoming weeks

**Comment by lankhaar** [comment]
I very recently bought myself a Eufy X10 Pro Omni, so now I'm here to join the rescue team of developers, hoping to speed up the process of integrating the device (and hopefully others) into HA.

I've read through the entire issue to get up to speed with all the findings of other devs. Thanks @CodeFoodPixels , @terabyte128 , @martijnpoppen and all others for your hard work so far!

@CodeFoodPixels , is this project still maintained (as in; do you merge any incoming PRs if I happen to make one)? And what's the status of the draft PR you made from the better dps branch? Is it somewhat stable and recommended to fork from, or is it recommended to fork from main branch?

@martijnpoppen, could you maybe elaborate a little bit on how to get the local key for a Eufy device? I used to have a Tuya vacuum before buying this one and getting the local key for that was simple through the Tuya developer portal, however since Eufy is in charge of that account in this case, I'm not sure how to get the local key in that case.

I wish y'all a lovely weekend and hope that together we can make some big progress in the near future on this topic!

**Comment by lee-b** [comment]
If I understand correctly, Eufy's server has no passwords per device, and may be piping any commands, from anyone, to those devices, which contain cameras, maps of building layouts, wifi credentials, etc.?

If so, this is unacceptable for me.  I sort of suspected as much and use my eufy without connectivity for now.

BUT, it should be possible to just set up an internal rabbitmq server and fake the DNS entry to point to it, right?  Does the Eufy check the server's cert?  Anyone tried something like this yet?

Even a proxy that whitelists known-safe commands would be a big improvement.

**Comment by hades200082** [comment]
@lee-b All very good points, but I don't see how they relate to getting it to work with Home Assistant?

Also, most of what you suggest would also remove the control of the robot from the Eufy app. 

One of the things I love about Home Assistant is that it makes it possible to create home automation without needing to remove the "normal" non-smart-home way that things work. i.e. A light switch still works as a light switch, but HA can also control it.

It sounds like you need a separate stack to handle what you need.

**Comment by lee-b** [comment]
@lee-b well, I was thinking that the implementation would be pretty lightweight, so I might think about adding it as an optional feature to home assistant -- private/local eufy support, rather than cloud. It wouldn't remove control from the eufy app if the phone is vpn'd into your local network and using local DNS; it would make that private too.

**Comment by martijnpoppen** [comment]
@lankhaar as said in my previous post 
The x10 doesn't have a local key. It doesn't work via Tuya anymore only via MQTT so there's not local control

**Comment by leeandy1** [comment]
I have been following this thread for some time in the hope of a solution for the X10. I know in an ideal world local control would be the desired outcome, but I am sure there are many users that would appreciate any type of Home Assistant integration for a start. I wish I had the skills to write an app or integration and could help with this. 

I'm sure I say it for many people in that we are grateful for everyone that is spending their own time to help on this. If it's testing that's needed I will be the first to put my hand up and help.

**Comment by lankhaar** [comment]
@martijnpoppen , thanks again for the elaborate response! I'm sorry for the confusion. I thought you meant that I don't need the local key to make an integration, but it wasn't clear initially that in fact there is no local key at all.

Are you aware of how Tuya/Eufy authenticate it's messages to the vacuum then?

**Comment by kmaid** [comment]
@martijnpoppen. I have been trying out your library and ported some of it to python. I can connect and list devices! I amusingly get all my notifications in german instead of english now 😆. Might not want to hardcode that

**Comment by martijnpoppen** [comment]
@kmaid yes still need to fix that. I got that from another repo on GitHub ;)

**Comment by leeandy1** [comment]
There has been some progress made in this repo https://github.com/arturonaredo/robovac

Using this fork I can get the X10 to show up in Home Assistant. No control or sensors but at least it gives some light at the end of the tunnel.

**Comment by cronner** [comment]
Wasn't mqtt figured out, can't it be added that way

**Comment by Biyadh00** [comment]
> There has been some progress made in this repo https://github.com/arturonaredo/robovac
> 
> Using this fork I can get the X10 to show up in Home Assistant. No control or sensors but at least it gives some light at the end of the tunnel.

Error
Config flow could not be loaded: {"message":"Invalid handler specified"}

**Comment by J2-Tech** [comment]
I am also getting the same error while attempting to configure the eufy robovac integration.

That being said, i made a very simple, very basic and probably not very secure REST API wrapper over @martijnpoppen 's work 
https://github.com/J2-Tech/eufy-clean-rest-wrapper/settings

It does not integrate as easily with home assistant, and in the end i'm not using it to automate my X10 because i am not skilled enough to understand how to de-compile the android application, reverse engineer the API and get the floor maps, schedules, etc. However, @martijnpoppen's SDK supports sending basic commands to the vacuum, so perhaps it could be a good reference

**Comment by CBDesignS** [comment]
> > There has been some progress made in this repo https://github.com/arturonaredo/robovac
> > Using this fork I can get the X10 to show up in Home Assistant. No control or sensors but at least it gives some light at the end of the tunnel.
> 
> Error Config flow could not be loaded: {"message":"Invalid handler specified"}

the logged error when debugging is enabled is :- 
Error occurred loading flow for integration robovac: No module named 'custom_components.robovac.tuyawebapi'

**Comment by jandrli** [comment]
Using Tuya's cloud solution is applicable if more direct methods have failed so far. Yes, in the case of my S1, it's pretty stripped down. Basically, I can just run a cleanup. Pause/stop/return don't work much, the other sensors are not available including the mode (I solved the automation dependent on it via the cleaning time sensor). However, at least it remembers the last map settings from the original app, which is enough for me.

**Comment by arbigator** [comment]
Some more progress being made here: [https://github.com/jeppesens/eufy-clean](url) by @jeppesens 

Works well so far for basic start/stop control.

**Comment by alondin** [comment]
Do you think that we will have control to s1 pro (rooms, zone, map etc) ever on ha?

**Comment by hades200082** [comment]
Not to sound ungrateful for the work people are doing for free to solve this or anything but I have a feeling that by the time we get the X10 Pro Omni working Eufy will have discontinued them and moved on to newer models :(

**Comment by MikeBP13** [comment]
At least here there is some progress:

> Some more progress being made here: [https://github.com/jeppesens/eufy-clean](url) by [@jeppesens](https://github.com/jeppesens)
> 
> Works well so far for basic start/stop control.

**Comment by Rbn3c** [comment]
The repository linked above received an update and works with individual rooms and scenes now. 

For me this is as close to full support as it gets without copying the Eufy app.

Install die Repo using HACS.

**Comment by gbcatrinoiu** [comment]
@Rbn3c mind explaining how to get it working? I just redownloaded but it does not seem to have any access to individual rooms.

**Comment by Rbn3c** [comment]
https://github.com/jeppesens/eufy-clean

The service call is as follows:

action: vacuum.send_command
target:
  entity_id: vacuum.robovac_x10_pro_omni
data:
  command: room_clean
  params:
    rooms:
      - 3
      - 4 

The README explains its use in more detail.

**Comment by hades200082** [comment]
Can it trigger a routine?

**Comment by ElCattiv0** [comment]
I just installed the latest version of eufy robovac on my Home Assistant, but after entering my credentials, it doesn't find my X10 Pro Omni. Instead the list of devices is empty and when I open up the configuration it just shows "Select the Vacuum to edit" and nothing more.
Any suggestions for newbies?

**Comment by CBDesignS** [comment]
> I just installed the latest version of eufy robovac on my Home Assistant, but after entering my credentials, it doesn't find my X10 Pro Omni. Instead the list of devices is empty and when I open up the configuration it just shows "Select the Vacuum to edit" and nothing more. Any suggestions for newbies?

Give this one a try. it seems to be further developed and supports the x10 pro, its basic but sort of working. 
https://github.com/jeppesens/eufy-clean

**Comment by ElCattiv0** [comment]
Works now. Thank you :)

**Comment by CavalliLane** [comment]
Happy to test if needed

**Comment by CavalliLane** [comment]
I would be happy to test eufy x10 pro also...

Really would be great to be able to include this in my dashboard...  my hubby is not tech savy, and one place for all tech in the home would be amazing!!!

Thanks in advance for your help with this!!

Jasen

**Comment by cronner** [comment]
dont expect this to be done never gonna happen I think

use https://github.com/jeppesens/eufy-clean and setup scenes that can be started from mqtt

**Comment by MikeBP13** [comment]
Since eufy is experimenting with different API for the X10 (REST?), is there a chance that this integration will support this model?

While the linked repository above does work, it is quirky, basic and lacks a lot of features...

**Comment by lankhaar** [comment]
@MikeBP13 ,

Where did you get that info from? I'm curious to read more about it!

**Comment by MikeBP13** [comment]
> [@MikeBP13](https://github.com/MikeBP13) ,
> 
> Where did you get that info from? I'm curious to read more about it!

https://github.com/jeppesens/eufy-clean/issues/43

**Comment by CBDesignS** [comment]
Rest was tried for about 1 whole week then they removed the back-end server api points and reverted to the original way again, they may return in the future but who knows. they have also seemed to have changed other things. cloud connection for devices. not sending full dps keys now.I was getting a whole host of invalid gtoken log lines and no connection to retrieve machine info, lots of changes happening under the covers. the 2 latest firmware updates I have recived to my x10 pro omni have chopped things up and moved things about. since the dps data is now only sent while the device is active and for a certain length of time after its docked/idle things seem to be a little more complicated. I am hunting for a wakeup command to send that might make things easier, as once its gone to sleep it only sends the machine status over mqtt dps.

**Comment by MikeBP13** [comment]
> Rest was tried for about 1 whole week then they removed the back-end server api points and reverted to the original way again, they may return in the future but who knows. they have also seemed to have changed other things. cloud connection for devices. not sending full dps keys now.I was getting a whole host of invalid gtoken log lines and no connection to retrieve machine info, lots of changes happening under the covers. the 2 latest firmware updates I have recived to my x10 pro omni have chopped things up and moved things about. since the dps data is now only sent while the device is active and for a certain length of time after its docked/idle things seem to be a little more complicated. I am hunting for a wakeup command to send that might make things easier, as once its gone to sleep it only sends the machine status over mqtt dps.

Thanks for the explanation. I hoped things will get easier 😞

**Comment by arwindersingh82** [comment]
Just bought the x10 pro... is there any chances of futurue support for the x10 pro or is it a return and get a different model?  I really need the mop to integrate to HA so i'm not sure if i should keep it!

**Comment by cronner** [comment]
> Just bought the x10 pro... is there any chances of futurue support for the x10 pro or is it a return and get a different model? I really need the mop to integrate to HA so i'm not sure if i should keep it!

If you look at the thread it's never gonna happen I think, using MQTT atm for scenes only not great

**Comment by AmyJeanes** [comment]
> Just bought the x10 pro... is there any chances of futurue support for the x10 pro or is it a return and get a different model? I really need the mop to integrate to HA so i'm not sure if i should keep it!

If a proper HA integration is a must I would look elsewhere unfortunately, at most you can send manually written MQTT messages to start a particular job and see it's status but there's no info on where it is within the house or also if it's coming back home to charge temporarily or finishing a job they both look the same. Pretty sure it's all cloud based too.
